internal/core/domain/trade_service.go

```diff
@@ -82,226 +86,187 @@ func (t *Trade) Accept(
 func (t *Trade) Complete(psetBase64 string) (*CompleteResult, error) {
-	if t.IsCompleted() {
+	if t.Status.Code >= Completed {
 		return &CompleteResult{OK: true, TxHex: t.TxHex, TxID: t.TxID}, nil
 	}
 
-	if !t.IsAccepted() {
-		return nil, ErrMustBeAccepted
-	}
-
-	if err := pkgswap.ValidateCompletePset(pkgswap.ValidateCompletePsetOpts{
-		PsetBase64:         psetBase64,
-		InputBlindingKeys:  t.SwapAcceptMessage().GetInputBlindingKey(),
-		OutputBlindingKeys: t.SwapAcceptMessage().GetOutputBlindingKey(),
-		SwapRequest:        t.SwapRequestMessage(),
-	}); err != nil {
-		t.Fail(
-			t.SwapAccept.ID,
-			FailedToCompleteStatus,
-			pkgswap.ErrCodeFailedToComplete,
-			err.Error(),
-		)
-		return &CompleteResult{OK: false}, nil
+	if t.IsExpired() {
+		if !t.Status.Failed {
+			t.Status.Failed = true
+		}
+		return nil, ErrTradeExpired
 	}
 
-	opts := wallet.FinalizeAndExtractTransactionOpts{
-		PsetBase64: psetBase64,
-	}
-	txHex, txHash, err := wallet.FinalizeAndExtractTransaction(opts)
-	if err != nil {
-		t.Fail(
-			t.SwapAccept.ID,
-			FailedToCompleteStatus,
-			pkgswap.ErrCodeFailedToComplete,
-			err.Error(),
-		)
-		return &CompleteResult{OK: false}, nil
+	if !t.IsAccepted() {
+		return nil, ErrTradeMustBeAccepted
 	}
 
-	swapCompleteID, swapCompleteMsg, err := pkgswap.Complete(pkgswap.CompleteOpts{
-		Message:    t.SwapAccept.Message,
-		PsetBase64: psetBase64,
-	})
+	swapCompleteID, swapCompleteMsg, err := SwapParser.SerializeComplete(
+		t.SwapRequestMessage(),
+		t.SwapAcceptMessage(),
+		psetBase64,
+	)
 	if err != nil {
 		t.Fail(
 			t.SwapAccept.ID,
-			FailedToCompleteStatus,
-			pkgswap.ErrCodeFailedToComplete,
+			err.Code,
 			err.Error(),
 		)
 		return &CompleteResult{OK: false}, nil
 	}
 
+	txHex, txHash, _ := PsetManager.GetTxHex(psetBase64)
 	t.SwapComplete.ID = swapCompleteID
 	t.SwapComplete.Message = swapCompleteMsg
+	t.Status = CompletedStatus
 	t.PsetBase64 = psetBase64
 	t.TxHex = txHex
 	return &CompleteResult{OK: true, TxHex: txHex, TxID: txHash}, nil
 }
 
-func (t *Trade) Settle(settlementTime uint64) error {
-	t.Status = CompletedStatus
-	return t.AddBlocktime(settlementTime)
-}
+// Settle brings the trade from the Completed to the Settled status, unsets the
+// expiration time and adds the timestamp of the settlement (it could be
+// a blocktime possibly) .
+func (t *Trade) Settle(settleTime uint64) (bool, error) {
+	if t.Status.Code == Settled {
+		return true, nil
+	}
 
-// Fail sets the status of the trade to the provided status and creates the
-// serialized SwapFail message and id for the given swap, errCode and errMsg
-func (t *Trade) Fail(swapID string, tradeStatus Status, errCode pkgswap.ErrCode, errMsg string) {
-	swapFailID, swapFailMsg, _ := pkgswap.Fail(pkgswap.FailOpts{
-		MessageID:  swapID,
-		ErrCode:    errCode,
-		ErrMessage: errMsg,
-	})
-	t.SwapFail.ID = swapFailID
-	t.SwapFail.Message = swapFailMsg
-	t.Status = tradeStatus
+	if t.IsExpired() {
```

Comment:
Here is tricky: we consider a swap completed (ie. with all the signatures) expired if for some reasons to be included in a block it takes more time, not dependent on either the user and the daemon?

Reply:
yeah, you're right.. if for some reason a completed trade exceeds the expiration time to be included in the blockchain it couldn't be marked as settled by the daemon. Shall we just remove this check then?

Reply:
Question: should we enforce that the provided settlemetTime (de facto a blocktime) is actually greater than the complete time?

Reply:
I would not enforce just because is not something that should block the trade: if it hapens something nasty with wither our machine clock or with the explorer block time, so lets favor trading instead of logical correctness

Reply:
I agree indeed. Already pushed the changes to check the expiration only in Complete() where the trade must be brought from Accepted to Completed before a certain expiration timestamp.